### PR TITLE
Daily calculation of user stats

### PR DESCRIPTION
**Pull request by crericha**

### CODE REVIEW

File: lib/bus.rb [hunk 1]
```diff
@@ -25,11 +34,15 @@ def self.[](key)
...
+    redis.pipelined do |pipeline|
+      pipeline.set(key, value)
+      pipeline.expire(key, ttl_for(key))
     end
 
     value
   end
+
+  def self.ttl_for(key)
+    TTL.find { |k, _v| key.include?(k) }&.last || 0
```
Comment (tobymao) [inline lib/bus.rb:46]: why not just TTL[key] is key some kind of dict?

File: lib/user_stats.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  K_NEW = 60.0
+  YEARS_INCLUDED = 4
+  BATCH_SIZE = 1000
+
+  def self.calculate_stats
+    user_stats = Hash.new { |hash, key| hash[key] = Hash.new([1200, 0]) }
+    on_or_after = Date.today - (365 * YEARS_INCLUDED)
+    offset = 0
+
+    loop do
```
Comment (tobymao) [inline lib/user_stats.rb:16]: i think there's a sequel method that let's you iterate in batches. checkout paged_each(rows_per_fetch: 1000)

File: lib/user_stats.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  def self.k_for(num_players, num_plays)
+    (num_plays <= 30 ? K_NEW : K) / (num_players - 1)
+  end
+
+  def self.update_stats(player, elo_change, category, stats)
+    old_elo, plays = stats[player][category]
+    stats[player][category] = [old_elo + elo_change[category], plays + 1]
+  end
+
+  def self.store_stats(user_id, stats)
```
Comment (tobymao) [inline lib/user_stats.rb:77]: this stores one at a time maybe add a pipeline on top so it's only one connection?

File: lib/bus.rb (hunk 1 above)
Comment (crericha) [reply lib/bus.rb:46]: Key has a unique id in it. Example, if the user's id is 1234, the key would be "USER_STATS:1234"

File: lib/user_stats.rb (hunk 2 above)
Comment (crericha) [reply lib/user_stats.rb]: Awesome. Fixed.

File: lib/user_stats.rb (hunk 3 above)
Comment (crericha) [reply lib/user_stats.rb]: Fixed.

File: lib/bus.rb (hunk 1 above)
Comment (tobymao) [reply lib/bus.rb:57]: this should be explicit, get the key by splitting the string and then doing normal hash lookip

Comment (crericha) [reply lib/bus.rb]: Fixed.

File: lib/user_stats.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+    Game.eager(:players).where(status: %w[finished archived], manually_ended: false)
+                        .where { finished_at >= on_or_after }
+                        .order(:finished_at)
+                        .paged_each(rows_per_fetch: BATCH_SIZE) do |game|
+      calculate_game_stats(game, user_stats)
+    end
+
+    store_stats(user_stats)
+    true
```
Comment (tobymao) [inline lib/user_stats.rb:23]: why return true?

File: lib/bus.rb [hunk 5]
```diff
@@ -24,12 +33,28 @@ def self.[](key)
...
 
+  def self.store_keys(keys)
+    redis.pipelined do |pipeline|
+      keys.each do |key, value|
+        pipeline.set(key, value)
+        pipeline.expire(key, ttl_for(key))
+      end
+    end
+
+    true
```
Comment (tobymao) [inline lib/bus.rb:44]: why return true?

File: lib/bus.rb [hunk 6]
```diff
@@ -7,7 +7,16 @@ module Bus
   TTL = 60 * 5
   URL = 'redis://redis:6379'
 
-  USER_TS = 'user_ts:%d'
+  USER_TS_PREFIX = 'user_ts'
```
Comment (tobymao) [inline lib/bus.rb:10]: freeze these as well

**Review by tobymao — APPROVED** [review]

File: lib/bus.rb [hunk 7]
```diff
@@ -24,12 +33,28 @@ def self.[](key)
...
+    redis.pipelined do |pipeline|
+      pipeline.set(key, value)
+      pipeline.expire(key, ttl_for(key))
     end
 
     value
   end
+
+  def self.ttl_for(key)
+    prefix = key.split(':').first
```
Comment (tobymao) [inline lib/bus.rb:57]: split(':')[0]

File: lib/user_stats.rb (hunk 4 above)
Comment (crericha) [reply lib/user_stats.rb:23]: Before it was returning the user_stats hash, which is long. What would you like it to return?

File: lib/bus.rb (hunk 5 above)
Comment (crericha) [reply lib/bus.rb:44]: It succeeded? Would you prefer nothing?

File: lib/user_stats.rb (hunk 4 above)
Comment (tobymao) [reply lib/user_stats.rb:23]: well if you don't need it to return anything don't return anything

File: lib/bus.rb (hunk 5 above)
Comment (tobymao) [reply lib/bus.rb:44]: do these methods except a return?

File: lib/bus.rb (hunk 7 above)
Comment (crericha) [reply lib/bus.rb:57]: Is [0] more efficient?

File: lib/user_stats.rb (hunk 4 above)
Comment (crericha) [reply lib/user_stats.rb:23]: So put a nil at the bottom?

File: lib/bus.rb (hunk 7 above)
Comment (tobymao) [reply lib/bus.rb:57]: no, but i prefer it because it's more concise

File: lib/user_stats.rb (hunk 4 above)
Comment (tobymao) [reply lib/user_stats.rb:23]: no, just don't do anything

Comment (crericha) [reply lib/user_stats.rb:23]: Ruby will return the output of the last line of the method. What's your opinion on what to do if that last line is a really big hash, string, etc.?

Comment (tobymao) [reply lib/user_stats.rb:23]: it's fine, just don't use it, you should only return stuff purposefully if you're going to use the value

**Pull request merged**